streaming/cleanup.go

```diff
@@ -0,0 +1,221 @@
+		var pubsubCfg internal.PubsubResources
+		err = json.Unmarshal([]byte(shardExecutionData.PubsubResources), &pubsubCfg)
+		if err != nil {
+			fmt.Printf("Unable to read Pubsub metadata for deletion\n")
+			return err
+		}
+		CleanupPubsubResources(ctx, pubsubClient, storageClient, pubsubCfg, project)
+	}
+	if options.Monitoring {
+		logger.Log.Debug("Attempting to delete shard monitoring resources...\n")
+		var monitoringResources internal.MonitoringResources
+		err := json.Unmarshal([]byte(shardExecutionData.MonitoringResources), &monitoringResources)
+		if err != nil {
+			logger.Log.Debug("Unable to read monitoring metadata for deletion\n")
+			return err
+		}
+		CleanupMonitoringDashboard(ctx, monitoringResources.DashboardName, project)
+	}
+	return nil
+}
+
+func GetJobDetails(ctx context.Context, migrationJobId string, dataShardIds []string, targetProfile profiles.TargetProfile, project string, instance string) (JobExecutionData, []ShardExecutionData, error) {
+	dbURI := fmt.Sprintf("projects/%s/instances/%s/databases/%s", project, instance, "spannermigrationtool_metadata")
+	client, err := utils.GetClient(ctx, dbURI)
+	if err != nil {
+		err = fmt.Errorf("can't create client for db %s: %v", dbURI, err)
+		return JobExecutionData{}, nil, err
+	}
+	defer client.Close()
+	txn := client.ReadOnlyTransaction()
+	defer txn.Close()
+
+	//fetch job level data
+	jobQuery := spanner.Statement{
+		SQL: fmt.Sprintf(`SELECT 
+								MigrationJobId,
+								SpannerDatabaseName,
+								TO_JSON_STRING(AggMonitoringResources) AS AggMonitoringResources,
+								IsShardedMigration,
+								CreateTimestamp
+							FROM JobExecutionData 
+							WHERE MigrationJobId = '%s'`, migrationJobId),
+	}
+	iter := txn.Query(ctx, jobQuery)
+	var jobExecutionData JobExecutionData
+	err = iter.Do(func(row *spanner.Row) error {
+		if err := row.ToStruct(&jobExecutionData); err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		err = fmt.Errorf("can't fetch job details from db for migration job: %s: %v", migrationJobId, err)
+		return JobExecutionData{}, nil, err
+	}
+
+	//fetch shard level data
+	shardQuery := spanner.Statement{
+		SQL: fmt.Sprintf(`SELECT 
+								MigrationJobId,
+								DataShardId,
+								TO_JSON_STRING(DataflowResources) AS DataflowResources,
+								TO_JSON_STRING(DatastreamResources) AS DatastreamResources,
+								TO_JSON_STRING(PubsubResources) AS PubsubResources,
+								TO_JSON_STRING(MonitoringResources) AS MonitoringResources,
+								CreateTimestamp
+							FROM ShardExecutionData 
+							WHERE MigrationJobId = '%s'`, migrationJobId),
+	}
+	iter = txn.Query(ctx, shardQuery)
+	shardExecutionDataList := []ShardExecutionData{}
+	for {
+		row, e := iter.Next()
+		if e == iterator.Done {
+			break
+		}
+		if e != nil {
+			err = e
+			break
+		}
+		var shardExecutionData ShardExecutionData
+		row.ToStruct(&shardExecutionData)
+		if filterbyDataShardId(shardExecutionData.DataShardId, dataShardIds) {
```

Comment:
Why not use the contains function in golang here?

Reply:
There is no contains for a list in golang

Reply:
https://pkg.go.dev/golang.org/x/exp/slices#Contains

Reply:
Ahh I see, I was trying  `dataShardIds.contains()`. Thanks will update

Reply:
Very interesting, the native `slices` package is available in Go 1.21 onwards. 
We are on go Go 1.19, and in order to consume this package, I will have to `go get` it. Seems like a lot for checking an array 😀